Demonstrate usage of foreign arrow function

hubertp opened the pull request:
### Pull Request Description

Including arrow language in the distribution by default. Added a basic example for creating an Arrow array.
Making sure that memory layout agrees with Arrow specification (padding, continuous allocation of memory chunks).
Related to #9118.

This should unblock work on allowing serialization/deserialization to/from Parquet but I'd like to delay it to a follow up ticket as it is going to be a significant amount of specialized work.

###  Important notes

As requested, Arrow array can currently be only modified via a builder pattern. Once the builder is sealed, no more modifications are allowed. At least not via official API. We all know that one can bypass those via a foreign function in JS or Python.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.

hubertp committed:
Demonstrate usage of foreign arrow function

Including arrow language in the distribution by default.
Added a basic example for creating an Arrow array.

Akirathan commented on build.sbt:
```diff
@@ -3021,7 +3020,8 @@ buildEngineDistribution := {
   updateLibraryManifests.value
   val modulesToCopy = componentModulesPaths.value.map(_.data)
-  val engineModules = Seq(file("runtime.jar"))
+  val arrow = Seq((`runtime-language-arrow`/Compile/packageBin ).value)
```
I think this is correct. Let's not assemble `runtime-language-arrow`, as that is unnecessary.

hubertp committed:
Missing Index_Out_Of_Bounds check

hubertp committed:
Show that one can write to Arrow Array

hubertp committed:
Missing file

JaroslavTulach commented on test/Base_Tests/src/Semantic/Arrow_Interop_Spec.enso:
```diff
@@ -0,0 +1,35 @@
+foreign arrow new_arrow = """
+    new[Int8]
+
+Vector.set self index elem =
```
There is `Polyglot.read_array_element` method, maybe you want to add `write_array_element` there.

hubertp committed:
Rounding logic that ensure correct padding

Arrow specification will enforce padding during serialization.
This change is mostly copying from existing Java reference
implementation.
Also while the padding is now correct, allocation of separate
`ByteDirect` buffers is not. TBC

hubertp committed:
Introduce Arrow_Array_Builder

The builder accepts an Arrow array and allows for populating the array
via the `set` method. Once the builder is `built` no more modifications
are allowed.

hubertp committed:
Ensure continous memory allocation

Arrow spec expects (informally) that memory allocated to data and
bitmaps is continous.
With this change, when allocating memory for some array, we allocate
full chunk of memory and then slice **from it** chunks for data and
bitmap, rather than allocating separate memory chunks.
TODO: add some reference counting in the future.

hubertp committed:
Fix tests

hubertp commented on test/Base_Tests/src/Semantic/Arrow_Interop_Spec.enso:
```diff
@@ -0,0 +1,41 @@
+            v = builder.build
+            v.set 8 10 . should_fail_with Unsupported_Argument_Types
+            v.at 8 . should_equal 127
+            builder.set 1 0 . should_fail_with Unsupported_Argument_Types
```
I'm not sure those are the right exceptions to throw. I'm open for suggestions (maybe we should introduce new kind of exception?).

radeusgd replied:
+1 for introducing a new exception.

The clearer and more specific the error messages are, the better.

JaroslavTulach replied:
OK. But (just for the record) my thumb down on yet another builtin (e.g. `InvalidOperation` builtin type).

jdunkerley commented on distribution/lib/Standard/Base/0.0.0-dev/src/Internal/Array_Like_Helpers.enso:
```diff
@@ -52,6 +52,9 @@ flatten array_like = @Builtin_Method "Array_Like_Helpers.flatten"
 insert_builtin : (Vector | Array) -> Integer -> (Vector | Array) -> Vector
 insert_builtin vector at items = @Builtin_Method "Array_Like_Helpers.insert_builtin"
 
+set_builtin : (Vector | Array) -> Integer -> Any -> Nothing
+set_builtin vector at item = @Builtin_Method "Array_Like_Helpers.set_builtin"
+
```
We don't need this anymore.

hubertp replied:
Yes, leftovers.

jdunkerley commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Arrow_Array_Builder.enso:
```diff
@@ -0,0 +1,44 @@
+
+@Builtin_Type
+type Arrow_Array_Builder
+    ## PRIVATE
+
+        Sets a provided value at a given index.
+    set : Integer -> Any -> Nothing ! Index_Out_Of_Bounds
+    set self index value =
+        Arrow_Array_Builder.set_builtin self index value
```
Normally, our builders only allow an "append" method where just set the current index and moves one further on.

hubertp replied:
added, left `set` in the `spec` as an extension. I can remove even that if you think it is a bad idea.

jdunkerley commented on distribution/lib/Standard/Base/0.0.0-dev/src/Polyglot.enso:
```diff
@@ -32,6 +32,15 @@ type Polyglot
     read_array_element : Any -> Integer -> Any
     read_array_element array index = @Builtin_Method "Polyglot.read_array_element"
 
+    ## PRIVATE
+       ADVANCED
+       Reads the element in a given polyglot array object.
+
+       Arguments:
+       - index: The index to get the element from.
+    write_array_element : Any -> Integer -> Any
+    write_array_element array index = @Builtin_Method "Polyglot.write_array_element"
+
```
Do we need this?

hubertp replied:
Added an example to the spec

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/interop/generic/WriteArrayElementNode.java:
```diff
@@ -0,0 +1,44 @@
+@BuiltinMethod(
+    type = "Polyglot",
+    name = "write_array_element",
+    description = "Write a value from the array specified by the index.",
```
```suggestion
    description = "Write a value to the array specified by the index.",
```

Do I misunderstand something or should this be `to` instead of `from`?

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/mutable/SetBuiltinVectorNode.java:
```diff
@@ -0,0 +1,56 @@
+    type = "Array_Like_Helpers",
+    name = "set_builtin",
+    description = "Sets a value into the Vector at the specified index.",
+    autoRegister = false)
```
I'm not sure we should be exposing this, our Vectors should be immutable.

Can we not encapsulate it to work just for Arrow arrays?

Also as @jdunkerley said, at least for current purposes we mostly only need `append`.

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrowArrayBuilder.java:
```diff
@@ -0,0 +1,242 @@
+
+  @ExportMessage
+  boolean isArrayElementInsertable(long index) {
+    throw CompilerDirectives.shouldNotReachHere();
```
Why we `shouldNotReachHere`?

hubertp replied:
It's overwritten by the concrete implementation

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrowArrayBuilder.java:
```diff
@@ -0,0 +1,242 @@
+    return true;
+  }
+
+  @ExportMessage
+  boolean isArrayElementReadable(long index) {
+    return true;
+  }
```
```suggestion
  @ExportMessage
  boolean isArrayElementReadable(long index) {
    return false;
  }
```

Shouldn't this be `false`?

I think `readArrayElement` should be consistent with `isArrayElementReadable` i.e. `readArrayElement(i)` may fail if and only if `isArrayElementReadable(i) == false`, and the exception to throw should be `UnsupportedOperationException`

radeusgd replied:
If we don't implement most of the Array API, should the ArrowArrayBuilder even say it is an array-like (`hasArrayElements = true`)?

IMO it should just be a helper builder that is _not_ an array. It should return the Arrow array and that probably should `hasArrayElements = true`.

But if the builder we barely even implement the array API and do not really rely on it, I'm not sure if we should pretend like it is an array.

IMHO it would be cleaner to have it be a distinct simple API (with `append` that we need).

But that is just my opinion and I may be missing something, so feel free to ignore.

hubertp replied:
delegates to interop

radeusgd commented on test/Base_Tests/src/Semantic/Arrow_Interop_Spec.enso:
```diff
@@ -0,0 +1,41 @@
+            builder.set 8 127
+            builder.set 8 128 . should_fail_with Unsupported_Argument_Types
+            v = builder.build
+            v.set 8 10 . should_fail_with Unsupported_Argument_Types
```
Why is `v.set` failing with Unsupported Argument Types?

should it be failing because of immutability?

hubertp replied:
> should it be failing because of immutability?

yes, that's by design

radeusgd commented on test/Base_Tests/src/Semantic/Arrow_Interop_Spec.enso:
```diff
@@ -0,0 +1,41 @@
+foreign arrow new_arrow = """
+    new[Int8]
+
+## Used only for testing immutability
+Vector.set self at item = set_builtin self at item
```
I'd prefer this to not be an extension function. Why not call `set_builtin` directly?

By being extension function it is unclear where we call the `Vector.set` method and when we call the method on another type. With direct call it would be more explicit.

JaroslavTulach replied:
Is this a left over? There is now the `Polyglot.write_array_element`. We should use the (mostly hidden) `write_array_element` function rather than advertising mutability directly on `Vector`.

hubertp replied:
> Is this a left over? 

yes

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Arrow_Array_Builder.enso:
```diff
@@ -0,0 +1,44 @@
+import project.Nothing.Nothing
+
+@Builtin_Type
+type Arrow_Array_Builder
```
Builtin type for a language which is not Enso? That's starting a maintenance mess.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/interop/generic/WriteArrayElementNode.java:
```diff
@@ -0,0 +1,44 @@
+import org.enso.interpreter.runtime.error.PanicException;
+
+@BuiltinMethod(
+    type = "Polyglot",
```
Having a builtin method in the engine for a `Polyglot` module is OK.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrowArrayBuilder.java:
```diff
@@ -0,0 +1,239 @@
+  @ExportMessage
+  Object readArrayElement(long index)
+      throws UnsupportedMessageException, InvalidArrayIndexException {
+    throw CompilerDirectives.shouldNotReachHere();
```
Please throw `UnsupportedMessageException` which is an expected interop exception instead of _panic exception_ like `shouldNotReachHere` which aborts execution altogether.

@radeusgd wrote:
> I think readArrayElement should be consistent with isArrayElementReadable

consistency is preferred, but there is no guarantee the callers of the `InteropLibrary` will honor the consistency! There is code in Truffle which doesn't bother checking `isXyz` and directly calls `xyz` catching the `UnsupportedException`. Such a use-case is considered valid.

hubertp replied:
In that case should [Vector](https://github.com/enso-org/enso/blob/develop/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/Vector.java#L49) also be changed? I tried to stick close to similar implementations.

JaroslavTulach replied:
Yes, it can be changed to not be a _misleading example_ to follow. However [Vector](https://github.com/enso-org/enso/blob/develop/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/Vector.java#L49) isn't the same case as you have here!

Please note that `Vector` is abstract shared class and all the real subclasses _override this message_ (but for example `Generic` does it with [completely different signature](https://github.com/enso-org/enso/blame/develop/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/Vector.java#L222) from Java perspective). The method in question should be `abstract` ideally, but then either `javac` or Truffle DSL complains when the method is made so.

Created https://github.com/enso-org/enso/pull/9312/commits/3a431a362be772703c542e704c5661e1cb08655e to consider some alternatives - #9312

JaroslavTulach commented on test/Base_Tests/src/Semantic/Arrow_Interop_Spec.enso:
```diff
@@ -20,14 +21,19 @@ add_specs suite_builder =
-            v.at 1 . should_equal 10
-            v.set 8 127
-            v.set 8 128 . should_fail_with Unsupported_Argument_Types
+            builder = Arrow_Array_Builder.from_arrow_array array_10
```
For a while I am asking myself, if this is the right pattern to follow? Shouldn't the `Vector.new_builder` and other builders be changed to receive a callback lambda function to perform the building in a clearly defined scope? E.g.
```ruby
v = Vector.new_builder builder->
    builder.set 8 127
    builder.set 1 10
```
it looks the similar and _feels more functional_. Moreover such a _lambda based approach_ is going to work fine in the IDE (unlike the current _unclosed_ builder).

As far as I can say the _lambda_ approach supports all use-cases except _creating builder and leaving it open_ - however that shall be an anti-pattern anyway, I believe/hope.

radeusgd replied:
@JaroslavTulach I love that proposal! I was worried that the current Vector Builder is very IDE unfriendly and will behave in weird ways. Your suggestion makes it safe. I'm all in for making this refactor, although I imagine it would be a separate PR as it's independent of Arrow.

hubertp replied:
I would prefer to make this change in a separate PR

JaroslavTulach replied:
This PR is trying to introduce _new builder_. It might be better to do the _new builder_ right (whatever _"right"_ means).

JaroslavTulach reviewed (changes requested):
We have designed, implemented and packaged the _Arrow language_ as a separate, Enso independent entity. This PR puts _Arrow Array Builder_ into `Standard.Base`. That's not right!

Arrow isn't a core concept of Enso. As such there should be _no builtins_ related to Arrow support.

Should the Arrow language remain independent from Enso language, then the builder has to be provided by the Arrow language, exposing its functionality via `InteropLibrary`. Simplest idea is to require two arguments to `new_arrow` constructor (created after parsing `new[Int8]` by the Arrow language) - the first being the size and the second an `isExecutable` `TruffleObject` (for example an Enso lambda function) that will be `execute(builder)` where `builder` is another `TruffleObject` reacting to `invokeMember("set")` & co.

Or we can decide that Arrow is natural part of Enso. But then we shouldn't have a separate _Arrow language_ JAR & co. Needless to say, I'd rather keep Enso small and have Arrow separate - unless there are inherent performance or UX limitations.

hubertp commented:
> We have designed, implemented and packaged the _Arrow language_ as a separate, Enso independent entity. This PR puts _Arrow Array Builder_ into `Standard.Base`. That's not right!

Despite the name, there is **nothing** Arrow-specific in that module.  And I think it would make sense to make it generic. I will update the PR.

hubertp committed:
PR comments

hubertp committed:
Make builder generic (in name)

hubertp commented:
@JaroslavTulach As I mentioned, there was nothing in the builder that made it arrow-specific. In fact this can serve as a generic polyglot array builder. The changes now reflect that.
But I also added a case where you can use Polyglot API to do the modifications in place. Then the sky is the limit (the potential damage as  well).

hubertp commented on test/Base_Tests/src/Semantic/Arrow_Interop_Spec.enso:
```diff
@@ -0,0 +1,41 @@
+foreign arrow new_arrow = """
+    new[Int8]
+
+Polyglot_Array_Builder.set self index value = Polyglot_Array_Builder.set_builtin self index value
```
I can remove this if people object. This is mostly for testing the functionality and it is not part of the official API

hubertp commented on test/Base_Tests/src/Semantic/Arrow_Interop_Spec.enso:
```diff
@@ -0,0 +1,41 @@
+            builder.set 1 0 . should_fail_with Unsupported_Argument_Types
+            Polyglot.write_array_element array_10 1 0
+            builder.at 1 . should_equal 0
+            v.at 1 . should_equal 0 # dangerous but allowed
```
@JaroslavTulach This demonstrates the potential consequences of using Polyglot API i.e. updating the builder after the fact that makes it "sealed". 
/cc @jdunkerley

JaroslavTulach replied:
> demonstrates the potential consequences of using Polyglot API 

This demonstrates that the _sealing only happens in the builder_, but not on the actual array. If we want object coming from our _arrow language_ to support sealing, why don't we put the `sealed` flag into `ArrowFixedArrayInt` & co.?

hubertp committed:
nit

hubertp committed:
More specific error

JaroslavTulach commented on engine/runtime-language-arrow/src/main/java/org/enso/interpreter/arrow/runtime/RoundingUtil.java:
```diff
@@ -0,0 +1,156 @@
+
+package org.enso.interpreter.arrow.runtime;
+
+public class RoundingUtil {
```
Does the class have to be public at all? Isn't it enough to be _package private_?

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/interop/generic/WriteArrayElementNode.java:
```diff
@@ -0,0 +1,44 @@
+      err.enter();
+      Builtins builtins = EnsoContext.get(this).getBuiltins();
+      throw new PanicException(builtins.error().makeInvalidArrayIndex(array, index), this);
+    } catch (UnsupportedTypeException e) {
```
`UnsupportedMessageException` leads to `makeTypeError` while `UnsupportedTypeException` leads to panic. Isn't that strange? Shouldn't `UnsupportedTypeException` lead to some _type error_?

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Error.java:
```diff
@@ -165,6 +168,10 @@ public Atom makeNoSuchConversion(Object target, Object that, UnresolvedConversio
     return noSuchConversion.newInstance(target, that, conversion);
   }
 
+  public Atom makeInvalidOperation(Object message) {
```
The `Object message` type is pretty generic and potentially wrong. We cannot pass `java.lang.String` there - that'd be an error. Recently I [changed one of such signatures](https://github.com/enso-org/enso/pull/9190/files#diff-859b6d81ecd37f1ded041958f50d39dfc3afffc2918cde8d6ab2b137ebfd5e25R198) to `Text`. Other alternative is `EnsoObject` or `TruffleObject` or overloaded methods supporting also `long` and `double`.

I can see the `Object` argument is accepted all around, but as I say, it is technically not correct.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/PolyglotArrayBuilder.java:
```diff
@@ -0,0 +1,269 @@
+    stdlibName = "Standard.Base.Data.Polyglot_Array_Builder.Polyglot_Array_Builder")
+public abstract class PolyglotArrayBuilder implements EnsoObject {
+
+  protected boolean sealed = false;
```
`protected`? Better to make (package) `private`.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/PolyglotArrayBuilder.java:
```diff
@@ -0,0 +1,269 @@
+
+  protected boolean sealed = false;
+
+  public void seal() {
```
`final`.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/PolyglotArrayBuilder.java:
```diff
@@ -0,0 +1,269 @@
+
+  @ExportMessage
+  boolean isArrayElementModifiable(long index) {
+    throw CompilerDirectives.shouldNotReachHere();
```
Please avoid `shouldNotReachHere` in favor of a correct _interop response_ (like `return false`, or not implementing the method or some `InteropException`).

JaroslavTulach commented on test/Base_Tests/src/Semantic/Arrow_Interop_Spec.enso:
```diff
@@ -0,0 +1,43 @@
+            v.at 8 . should_equal 127
+            builder.append 21 . should_fail_with Invalid_Operation
+            builder.set 1 21 . should_fail_with Invalid_Operation
+            Polyglot.write_array_element array_10 1 0
```
Let me recap what we achieved so far:
- we expose `new[Int8]` to public
- we let public to create `array_10` and hold it indefinitly
- we let public to modify the array _any time_ with `Polyglot.write_array_element array_10`
- on top of that we created some _fake facade_ called `Polyglot_Array_Builder` 
- the facade provides _no additional functionality_
- every operation can already be done on `array_10` by other means

I guess I am missing the point of the whole _builder exercise_. Whatever it tried to achieve (probably some _encapsulation_ or _giving the creator_ more rights) clearly isn't achieved. Btw. _Give the Creator of an Object More Rights_ pattern is described at page 79 of the _Practical API Design_ book.

JaroslavTulach reviewed (changes requested):
I continue to have an API design architectural problem with this PR.

> was nothing in the builder that made it arrow-specific. In fact this can serve as a generic polyglot array builder. The changes now reflect that.

Whatever was @jdunkerley reason for requesting the builder, it is not achieved as the usage of the new builder is completely optional and there are other means to achieve the same (and more) without the builder.

Proper API design shall start with description of the use-case. As a reviewer I am not supposed to guess the use-case. However I have to: If the use-case was _encapsulation of the write operation_, then it is not achieved! If the use-case is something else, then please mention it in the **Pull Request Description** - as currently the justification for majority of the code (I mean the _bloat related to all the new builder builtins_) in this PR is missing (or at last I cannot find it).

> But I also added a case where you can use Polyglot API to do the modifications in place. Then the sky is the limit (the potential damage as well).

Right. So what is your use-case related to `Polyglot.write_array_element`? Should one be allowed to _do the modifications in place_ from Enso language or not? The original decision when working on #5011 and related PRs was that we **don't want to expose write operations** in Enso.

Using _builder pattern_ is a good way to avoid _exposing write operations_, however it has to be done properly - e.g. (in current architectural design) the _builder has to be provided by the Arrow language_.

JaroslavTulach commented:
Assuming @jdunkerley main requirement is the: _don't let mutable arrays float thru the computation_, then ...

> .... Simplest idea is to ....

Change the syntax of the Arrow language to create the builder itself:
```ruby
foreign arrow new_array_int8 = """
    new[Int8]

b10 = new_array_int8 10
b10.append 2 
b10.append 4
arr = b10.build # or to_array?

arr.length . should_equal 10
arr.at 0 . should_equal 2
arr.at 1 . should_equal 4
arr.at 5 . is_nothing . should_be_true # maybe?

Polyglot.write_array_element arr 6 33 # raises an exception as the arr is already immutable
```
if this is what we want, then it all can be implemented in the _Arrow language_ itself without any need for so many builtins in the Enso engine. As a result of the changes the [test cases that use new](https://github.com/enso-org/enso/pull/8512/files#diff-0aeb0af4012fe400ce70effe933d40aee88497c73506bd96d3f389ed2081afb1R58) would need to be updated to _use the builder_ (newly returned from the factory method).

#### Arrow Array is already Enso Array

Please note that one can apply all Enso Array methods on the `arr` produced by _Arrow language_:
```ruby
b3 = new_array_int8 3
b3.append 1 
b3.append 2
b3.append 3
arr = b3.build 

v1 = arr.map _.to_text # yields ['1', '2', '3']
v2 = arr.fold 10 (+) # yields 16
v3 = Meta.type_of arr # yields Array
```

hubertp committed:
Push builder into arrow language

Avoiding the introduction of various builtins by pushing builder pattern
into allow language.

hubertp committed:
Merge branch 'develop' into wip/hubert/9118-arrow

hubertp committed:
nit

JaroslavTulach commented on engine/runtime-language-arrow/src/main/java/org/enso/interpreter/arrow/runtime/ArrowFixedArrayDate.java:
```diff
@@ -222,37 +162,11 @@ private static ZonedDateTime zonedDateTimeFromSeconds(long seconds, long nano, Z
 
-  private static int typeAdjustedIndex(long index, SizeInBytes unit) {
-    return Math.toIntExact(index * unit.sizeInBytes());
+  public static int typeAdjustedIndex(long index, int daySizeInBytes) {
```
Package private might be enough. Whole language seems to be in a single package (except `ArrowParser`).

JaroslavTulach commented on engine/runtime-language-arrow/src/main/java/org/enso/interpreter/arrow/runtime/ArrowFixedSizeArrayBuilder.java:
```diff
@@ -0,0 +1,353 @@
+
+  @ExportMessage
+  public boolean hasArrayElements() {
+    return true;
```
The _"builder"_ is in fact also an (Enso) `Array` and it can be mutated until somebody calls `build` on it. I'd be interested to see a _use-cases_ for such a design decision...

JaroslavTulach commented on test/Base_Tests/src/Semantic/Arrow_Interop_Spec.enso:
```diff
@@ -0,0 +1,40 @@
+            builder.at 10 . should_fail_with Index_Out_Of_Bounds
+            builder.append 2
+            builder.append 10
+            builder.at 1 . should_equal 10
```
The value at index 1 is 10 while few lines above it was `Nothing`. That's OK - I am just not sure why `builder` should have `at` method at all?

Moreover `Meta.type_of builder` is Enso `Array` - e.g. one can do `.map` or `.fold` on the `builder` and in general use it as an array.

JaroslavTulach commented on engine/runtime-language-arrow/src/main/java/org/enso/interpreter/arrow/runtime/ArrowFixedSizeArrayBuilder.java:
```diff
@@ -0,0 +1,353 @@
+  @ImportStatic(LogicalLayout.class)
+  static class InvokeMember {
+    @Specialization(guards = "isDateLayout(receiver.getUnit())")
+    public static Object doInvokeDate(
```
`doInvokeDate` seems like an left over.... but it is not! This means that the messages creates `ArrayFixedArrayDate` at the end.

JaroslavTulach commented on engine/runtime-language-arrow/src/main/java/org/enso/interpreter/arrow/runtime/ArrowFixedSizeArrayBuilder.java:
```diff
@@ -0,0 +1,353 @@
+      switch (name) {
+        case "build":
+          receiver.sealed = true;
+          return new ArrowFixedArrayInt(receiver.buffer, receiver.unit);
```
Is `doInvokeDate` and `doInvokeInt` the same except this line? Then it'd be better to avoid the duplicated code for the sake of future maintenance.

JaroslavTulach commented on engine/runtime-language-arrow/src/main/java/org/enso/interpreter/arrow/runtime/ArrowFixedSizeArrayBuilder.java:
```diff
@@ -0,0 +1,353 @@
+        @Cached.Shared("interop") @CachedLibrary(limit = "1") InteropLibrary iop)
+        throws UnsupportedMessageException {
+      if (receiver.sealed) {
+        throw UnsupportedMessageException.create();
```
The [writeArrayElement message](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/InteropLibrary.html#writeArrayElement-java.lang.Object-long-java.lang.Object-) can return three different `InteropException`s:
- [UnsupportedMessageException](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/UnsupportedMessageException.html)
- [UnsupportedTypeException](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/UnsupportedTypeException.html)
- [InvalidArrayIndexException](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/InvalidArrayIndexException.html)

If you throw better interop exceptions, the Enso users may get better error message on the Enso level.

Here I suggest to report `UnsupportedTypeException` with a `hint` saying the `builder` is already sealed.

hubertp replied:
It doesn't feel like `UnsupportedTypeException` is the right one, though. It's the problem with the receiver not with arguments.

JaroslavTulach replied:
The goal would be to improve the error message - getting `Not_Invokable` is misleading. Maybe you can return `false` from [isArrayElementModifiable](https://github.com/enso-org/enso/pull/9150/files#diff-2ad45a930106af4f2ab4cfd869e2f5fcaf07ce06a35721835b02a96ad3e8335bR341) when _sealed_ or throw your own [truffle exception](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/exception/AbstractTruffleException.html), but `UnsupportedMessageException` certainly has different meaning than _already sealed_.

>  problem with the receiver not with arguments.

Well, the _receiver changed type_ - it used to be mutable and now it is not mutable. I don't have a problem using simplest solution if it improves the error message even if it requires a bit of mental twist.

JaroslavTulach commented on engine/runtime-language-arrow/src/main/java/org/enso/interpreter/arrow/runtime/ArrowFixedSizeArrayBuilder.java:
```diff
@@ -0,0 +1,353 @@
+        throw UnsupportedMessageException.create();
+      }
+      if (!iop.isDate(value)) {
+        throw UnsupportedMessageException.create();
```
This should clearly be `UnsupportedTypeException`.

hubertp replied:
Agreed

JaroslavTulach commented on engine/runtime-language-arrow/src/main/java/org/enso/interpreter/arrow/runtime/ArrowFixedSizeArrayBuilder.java:
```diff
@@ -0,0 +1,353 @@
+        throw UnsupportedMessageException.create();
+      }
+      receiver.buffer.putLong(typeAdjustedIndex(index, receiver.unit), (iop.asLong(value)));
+    }
```
I don't see any `@Fallback` or catch-all `@Specialization`. What happens if you try to write wrong value type into wrong array type? Will it yield user friendly `UnsupportedTypeException` or will it crash on missing specialization?

JaroslavTulach commented on engine/runtime-language-arrow/src/main/java/org/enso/interpreter/arrow/runtime/ArrowFixedSizeArrayBuilder.java:
```diff
@@ -0,0 +1,353 @@
+    public static void doLong(
+        ArrowFixedSizeArrayBuilder receiver,
+        long index,
+        Object value,
```
Just make it `long value` you don't need `iop`.... but then you need the `@Fallback` specialization to report error nicely.

JaroslavTulach commented on engine/runtime-language-arrow/src/main/java/org/enso/interpreter/arrow/runtime/ArrowFixedSizeArrayBuilder.java:
```diff
@@ -0,0 +1,353 @@
+        throws UnsupportedMessageException, InvalidArrayIndexException {
+      validAccess(receiver, index);
+      if (!iop.fitsInLong(value)) {
+        throw UnsupportedMessageException.create();
```
As in many other cases, this should not be `UnsupportedMessageException`, but [UnsupportedTypeException](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/UnsupportedTypeException.html) with some `hint`.

JaroslavTulach commented on test/Base_Tests/src/Semantic/Arrow_Interop_Spec.enso:
```diff
@@ -0,0 +1,40 @@
+            builder.append 127
+
+            # Check that the value that will not fit into a byte is rejected
+            Test.expect_panic_with (builder.append 128) Not_Invokable
```
If you throw [UnsupportedTypeException](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/UnsupportedTypeException.html) in the `builder.writeArrayElement` interop message, then (I hope) you'll get better Enso exception than `Not_Invokable` here.

hubertp replied:
Nope. Invoke can't throw `UnsupportedTypeException` so I have to wrap it in `UnsupportedMessageException` anyway.

JaroslavTulach replied:
I believe there is [UnsupportedTypeException](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/UnsupportedTypeException.html) in the signature of [invokeMember](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/InteropLibrary.html#invokeMember-java.lang.Object-java.lang.String-java.lang.Object...-)

JaroslavTulach replied:
At https://github.com/enso-org/enso/pull/9150/files#diff-ebe312b126e000670288fc461a3e77fa0551b8e32341d54643d0a5e494f7446bR31 I see:
```
 Test.expect_panic_with (builder.append 128) Unsupported_Argument_Types
```
that's a reasonable exception type, thank you. However at https://github.com/enso-org/enso/pull/9150/files#diff-ebe312b126e000670288fc461a3e77fa0551b8e32341d54643d0a5e494f7446bR34 I can still find:
```
Test.expect_panic_with (builder.append 21) No_Such_Method
``` 
I find that misleading.

JaroslavTulach reviewed (approved):
From an **architecture perspective** the change is looking **good** now. It is located in the right place (e.g. `runtime-language-arrow`), it doesn't require any changes to `engine/runtime`. It fits perfectly the vision why _Arrow language_ was founded as a (loosely) separated project.

I have doubts about the decision to make `builder : Array`. Possibly that's unnecessary leakage of mutability into the Enso world. However that may stay as it is and remain a subject of future discussions.

I believe we should improve error reporting by using better subclasses of [InteropException](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/InteropException.html). Given the _typed array nature_ of Arrow buffers, I'd use [UnsupportedTypeException](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/UnsupportedTypeException.html) more.

I haven't verified the change from the point of performance. I assume there will be some benchmarking work in the future.

hubertp committed:
PR review

hubertp committed:
nits
